yaml/shared/src/main/scala/org/virtuslab/yaml/YamlDecoder.scala

```diff
@@ -12,64 +12,96 @@ import org.virtuslab.yaml.Node.*
+    value.toDoubleOption.toRight(cannotParse(value, "Double", s))
   }
 
-  given YamlDecoder[Float] = YamlDecoder { case ScalarNode(value, _) =>
-    Try(value.toFloat).toEither
+  given YamlDecoder[Float] = YamlDecoder { case s @ ScalarNode(value, _) =>
+    value.toFloatOption.toRight(cannotParse(value, "Float", s))
   }
 
-  given YamlDecoder[Short] = YamlDecoder { case ScalarNode(value, _) =>
-    Try(value.toShort).toEither
+  given YamlDecoder[Short] = YamlDecoder { case s @ ScalarNode(value, _) =>
+    value.toShortOption.toRight(cannotParse(value, "Short", s))
   }
 
-  given YamlDecoder[Byte] = YamlDecoder { case ScalarNode(value, _) =>
-    Try(value.toByte).toEither
+  given YamlDecoder[Byte] = YamlDecoder { case s @ ScalarNode(value, _) =>
+    value.toByteOption.toRight(cannotParse(value, "Byte", s))
   }
 
-  given YamlDecoder[Boolean] = YamlDecoder { case ScalarNode(value, _) =>
-    Try(value.toBoolean).toEither
+  given YamlDecoder[Boolean] = YamlDecoder { case s @ ScalarNode(value, _) =>
+    value.toBooleanOption.toRight(cannotParse(value, "Boolean", s))
+  }
+
+  given YamlDecoder[Any] = new YamlDecoder {
+    def construct(node: Node)(using settings: LoadSettings = LoadSettings.empty) = {
+      node match {
+        case ScalarNode(value, tag) if Tag.primitives.contains(tag) =>
+          tag match {
+            case Tag.nullTag => Right(None)
+            case Tag.boolean => value.toBooleanOption.toRight(cannotParse(value, "Boolean", node))
+            case Tag.int =>
+              if value.startsWith("0b") then
+                Try(Integer.parseInt(value.drop(2), 8)).toEither.swap
+                  .map(t => ConstructError.from(t, "Int", node))
+                  .swap
+              else if value.startsWith("0x") then
+                Try(Integer.parseInt(value.drop(2), 8)).toEither.swap
+                  .map(t => ConstructError.from(t, "Int", node))
+                  .swap
+              else value.toIntOption.toRight(cannotParse(value, "Int", node))
+            case Tag.float =>
+              value.toDoubleOption.toRight(cannotParse(value, "Double", node))
+            case Tag.str => Right(value)
+          }
+        case MappingNode(mappings, Tag.map) =>
+          val decoder = summon[YamlDecoder[Map[Any, Any]]]
+          decoder.construct(node)
+        case SequenceNode(seq, Tag.seq) =>
+          val decoder = summon[YamlDecoder[Seq[Any]]]
+          decoder.construct(node)
+        case _ =>
+          settings.constructors.get(node.tag) match {
+            case Some(decoder) => decoder.construct(node)
+            case None => Left(ConstructError(s"Could't create construct instance for ${node.tag}"))
+          }
+      }
+    }
   }
 
   given [T](using c: YamlDecoder[T]): YamlDecoder[Option[T]] = YamlDecoder {
-    case ScalarNode(value, _) =>
+    case ScalarNode(value, tag) =>
       value match
-        case "null" | "" => Right(None)
+        case _ if tag == Tag.nullTag => Right(None)
```

Comment:
I don't think, we should remove `case "null" | ""`, we should allow to specify `null` with "null" or ""

Reply:
but `""` is an empty string, not a `null` value, we don't differentiate this but we should. Maybe in `ParserImpl#parseNode` we should alter the last case to always emit `null` tag with an empty string? Thanks to that we would be able to recognize explicit empty string `("", tag:yaml.org,2002:str)` (because we parsed Scalar token and the tag will resolve to string) from `("", tag:yaml.org,2002:null)` (because we expect one of the node's tokens, but we didn't encounter any, so we will explicitly add null tag)

Reply:
It's make sense, I approve.